tempesta_fw/sock_srv.c

```diff
@@ -1248,262 +1699,521 @@ static int
+	 * - must delete all sg_cfg_list->parsed_sg
+	 * - must delete all active servers and groups (server.c:sg_list).
+	 */
+	tfw_sg_drop_reconfig();
+	tfw_cfgop_cleanup_srv_cfgs(true);
 
-	tfw_sock_srv_delete_all_conns();
-	tfw_sg_release_all();
+	/*
+	 * Active configuration will be cleaned up if tfw_sock_srv_stop()
+	 * was called.
+	 */
 }
 
+/**
+ * Parse graceful shutdown time.
+ */
 static int
-tfw_sock_srv_start(void)
+tfw_cfgop_grace_time(TfwCfgSpec *cs, TfwCfgEntry *ce)
 {
-	int ret;
+	return tfw_cfgop_intval(cs, ce, &tfw_cfg_grace_time_reconfig);
+}
 
+static int
+tfw_sock_srv_cfgstart(void)
+{
+	INIT_LIST_HEAD(&sg_cfg_list);
+	if (!(tfw_cfg_sg_opts = __tfw_cfgop_new_sg_cfg(TFW_CFG_SG_OPTS_NAME)))
+		return -ENOMEM;
+	if (!(tfw_cfg_sg_def = tfw_cfgop_new_sg_cfg_def())) {
+		tfw_cfgop_cleanup_srv_cfg(tfw_cfg_sg_opts, true);
+		return -ENOMEM;
+	}
+	tfw_cfg_sg = tfw_cfg_sg_opts;
+	tfw_cfg_use_sticky_sess = false;
+	memset(&tfw_cfg_is_set, 0, sizeof(tfw_cfg_is_set));
+
+	return 0;
+}
+
+static int
+tfw_sock_srv_cfgend(void)
+{
+	int r;
 	/*
-	 * The group "default" is created implicitly, and only when
-	 * a server outside of any group is found in the configuration.
+	 * The group 'default' to be created implicitly if at least one server
+	 * is defined outside of any group and there is no explicit 'default'
+	 * group.
 	 */
-	if (tfw_cfg_out_slstsz) {
-		tfw_cfg_out_sg = tfw_sg_new("default", GFP_KERNEL);
-		if (!tfw_cfg_out_sg) {
-			TFW_ERR_NL("Unable to add default server group\n");
-			return -EINVAL;
+	if (!tfw_cfg_sg_def)
+		return 0;
+	if (!tfw_cfg_sg_def->parsed_sg->srv_n) {
+		tfw_cfgop_cleanup_srv_cfg(tfw_cfg_sg_def, true);
+		tfw_cfg_sg_def = NULL;
+		return 0;
+	}
+	/* Options for implicit group are not filled, use current defaults. */
+	tfw_cfgop_sg_copy_opts(tfw_cfg_sg_def->parsed_sg,
+			       tfw_cfg_sg_opts->parsed_sg);
+	tfw_cfgop_sg_copy_sched_arg(&tfw_cfg_sg_def->sched_arg,
+				    tfw_cfg_sg_opts->sched_arg);
+	tfw_cfg_sg_def->parsed_sg->sched = tfw_cfg_sg_opts->parsed_sg->sched;
+	tfw_cfg_sg_def->nip_flags = tfw_cfg_sg_opts->nip_flags;
+	tfw_cfg_sg_def->sticky_flags = tfw_cfg_sg_opts->sticky_flags;
+	tfw_cfg_sg_def->sched_flags = tfw_cfg_sg_opts->sched_flags;
+
+	if ((r = tfw_cfgop_setup_srv_group(tfw_cfg_sg_def)))
+		return r;
+	tfw_cfg_sg_def = NULL;
+
+	return 0;
+}
+
+static int
+tfw_cfgop_update_srv(TfwServer *orig_srv, TfwCfgSrvGroup *sg_cfg)
+{
+	TfwServer *srv;
+	int r;
+
+	if (!(srv = tfw_server_lookup(sg_cfg->parsed_sg, &orig_srv->addr)))
+		return -EINVAL;
+
+	TFW_DBG_ADDR("Update server options", &srv->addr);
+
+	orig_srv->weight = srv->weight;
+
+	if (orig_srv->conn_n < srv->conn_n) {
+		r = tfw_sock_srv_append_conns_n(srv,
+						srv->conn_n - orig_srv->conn_n);
+		if (r)
+			return r;
+		orig_srv->conn_n = srv->conn_n;
+	}
+	else if (orig_srv->conn_n > srv->conn_n) {
+		/*
+		 * TODO: shrink number of connections. Disconnects are performed
+		 * asynchronously, can't destroy connection here and now.
```

Comment:
When we need to shrink number of connections to a given server, we must do two things: disconnect and destroy some connections; and update scheduler data for that server. The scheduler data is normally common for the whole server group. 

Disconnecting of connections is an asynchronous operation. So we have to keep connections in `srv->conn_list` until disconnect is completed. In the same time right after scheduling for disconnects we use all connections from `srv->conn_list` to build scheduler data. Dead connections poison scheduler, since number of scheduling attempts is limited.

UPD: Somehow i missed a possible solution: we can check `TFW_CONN_B_DEL` flag when building scheduler data to skip connections scheduled to disconnect. Any way that can be done as a separate task.